src/mpi/request/request_impl.c

```diff
@@ -855,28 +865,8 @@ int MPIR_Wait(MPIR_Request * request_ptr, MPI_Status * status)
     int mpi_errno = MPI_SUCCESS;
     int active_flag;
 
-    if (!MPIR_Request_is_complete(request_ptr)) {
-        /* If this is an anysource request including a communicator with
-         * anysource disabled, convert the call to an MPI_Test instead so we
-         * don't get stuck in the progress engine. */
-        if (unlikely(MPIR_Request_is_anysrc_mismatched(request_ptr))) {
-            mpi_errno = MPIR_Test(request_ptr, &active_flag, status);
-            goto fn_exit;
-        }
-
-        if (MPIR_Request_has_poll_fn(request_ptr)) {
-            while (!MPIR_Request_is_complete(request_ptr)) {
-                mpi_errno = MPIR_Grequest_poll(request_ptr, status);
-                MPIR_ERR_CHECK(mpi_errno);
-
-                /* Avoid blocking other threads since I am inside an infinite loop */
-                MPID_THREAD_CS_YIELD(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
-            }
-        } else {
-            mpi_errno = MPID_Wait(request_ptr, status);
-            MPIR_ERR_CHECK(mpi_errno);
-        }
-    }
+    mpi_errno = MPID_Wait(request_ptr, status);
+    MPIR_ERR_CHECK(mpi_errno);
```

Comment:
I am looking the commit "request: Move progress logic for MPIR_Test/Wait". The code here is to check whether it is a general request or we need to convert "wait" into "test". Both are logically above the device layer. So I am not sure about the merit of this commit.

I am a bit confused that at this point, all reference to `MPIR_Wait` seems gone; then we must have broken the general request polling at some previous commits, right?

Reply:
I was a bit confused too. My reasoning was that all other generalized request handling is done in MPIR_*_state functions from what I could gather from the code. I guess the alternative is move all the grequest handling up a layer?

My other thought is: is this just an artifact of single request vs. multi-request APIs? Are we trying to avoid multiple loops over the request array and so consolidate the logic in a single place for multi-request APIs? Otherwise we could easily run some simple loops in the upper layers to handle non-device specific logic...

Reply:
>  I guess the alternative is move all the grequest handling up a layer?

I would like that. `MPIR_*_state` are concerned with lower layer details such as vcis (or whatever in the "progress state"). So moving grequest out of `MPIR_*_state` makes sense.

Reply:
OK. I changed so grequest polling is done at the MPIR layer. Obviously this only addresses single request test/wait at the moment, but should be a good start.